mysql-test/suite/rpl/t/rpl_mts_spco_deadlock_slave_trans_retries_assertion.test

```diff
@@ -0,0 +1,159 @@
+# ==== Purpose ====
+#
+# This test verifies that deadlock errors arising due to lock conflicts are
+# handled properly when --slave-preserve-commit-order is enabled on the replica
+# server.
+#
+# ==== Implementation ====
+#
+# 1. Create source-replica topolgy.
+# 2. Setup testing environment.
+#    2.1. Create table on source and insert few rows.
+#    2.2. Setup necessary variables on replica server.
+# 3. Execute transactions on source server that would generate deadlock on the
+#    replica server.
+#
+#    The commands used to generate deadlock in this test are:
+#
+#    INSERT INTO t1 (11, NULL);
+#    DELETE FROM t1 WHERE c2 <= 3;
+#
+#    With the help of debug points on the source server, simulate that they are
+#    executed parallely on master and make sure INSERT is binlogged before
+#    DELETE.
+#
+#    On replica server, they are applied parallel by w1 and w2. And the deadlock is:
+#      w1                                        w2
+#    -------------------------                 --------------------------
+#     applying INSERT                          applying DELETE
+#                                              get the row locks first.
+#     waiting for w1 to release
+#     the row lock.
+#                                              waiting for w2 to commit
+#
+#     For these two special statements, INSERT doesn't block DELETE, but DELETE
+#     does block the INSERT because of the gap locks.
+#
+# 4. When the replica server finds the deadlock, it will rollback w2 in order
+#    to allow w1 to commit first. When w2 rolls back the transaction and retries
+#    it, simulate that all retries are exhausted so that it tries to
+#    log a message in the error log.
+# 5. Wait till the co-ordinator thread to error out with ER_LOCK_DEADLOCK.
+# 6. Restart replication threads, sync the replica and verify that table data is
+#    consistent.
+# 7. Cleanup.
+#
+# ==== References ====
+#
+# PS-7231 Assertion failure `m_status == DA_ERROR` in `Diagnostics_area::mysql_errno()`
+# PS-7197 Multi-threaded slave randomly stacks
```

Comment:
7197 has any role here?
also I guess the title should have randomly hangs instead of "stacks".. may be they meant stuck.

Reply:
I mentioned PS-7197 because PS-7231 is marked as a subtask of PS-7197 and I was not sure about if I should mention only the base bug or the subtask.

And regarding the title, it should have been `stuck` and not `stack`. Will update the JIRA title now.

Reply:
I mentioned PS-7197 because PS-7231 is marked as a subtask of PS-7197 and I was not sure about if I should mention only the base bug or the subtask.

And regarding the title, it should have been `stuck` and not `stack`. I will update the JIRA title now.

Reply:
Changed it to `PS-7197 Multi-threaded slave hangs when slave_trans_retires gets exhausted`